### ISSUE TITLE
Starting REPL fails for leiningen projects with more than one project.clj

### ISSUE THREAD

**Issue by marcomorain** [BetterThanTomorrow/calva#871]
One of our projects has multiple files named `project.clj`:

```
marc@blaster ~/dev/circleci/circle $ find . -name project.clj
./project.clj
./src/circle/project.clj
./src/circle/http/api/v1/v1_0/project.clj
./src/circle/http/api/v1/v1_1/project.clj
./src/circle/http/api/v1/shared/project.clj
./src/circle/http/api/v2/insights/project.clj
./src/circle/http/api/v2/project.clj
./src/circle/http/api/v2/entities/project.clj
./src/circle/model/project.clj
./src/circle/model/api/project.clj
```

Only the first of these files is a leiningen project - the others are just namespaces.

The only way to start a REPL in this project is to open `project.clj` in the root directory and run the command to start a REPL. If you try to start a REPL with any file open in `src/circle/`, starting the REPL fails.

![image](https://user-images.githubusercontent.com/448001/102069867-3d9ed280-3df6-11eb-9686-2e575e097a34.png)

The problem seems to be that Calva tries to find a `project.clj` starting in the `pwd` and moving up to the root of the workspace. If it finds a file on the way named `project.clj`, I guess it tries to parse it as a lein project, and if that fails, then Calva gives up on trying to start the REPL.

**Comment by bpringe** [comment]
> The problem seems to be that Calva tries to find a project.clj starting in the pwd and moving up to the root of the workspace.

Besides perhaps the way it's failing, this is by design. This way people can have multi project directories, and Calva will start at a project repl at the level of the file that is open.

> Only the first of these files is a leiningen project - the others are just namespaces.

Can you elaborate here? Do you mean those other project.clj files are being used as regular namespace/code files (ending with `.project`? I'd imagine that would be considered an anti-pattern using lein.

**Comment by bpringe** [comment]
I think the best thing to do here would be to rename those project namespaces and files to something else. I'm not sure we want to account for them being anything besides a lein poject file. However, I think that error message we throw could be more descriptive and inform the user that it must be a lein project file.

@PEZ, what do you think?

**Comment by marcomorain** [comment]
> Can you elaborate here? Do you mean those other project.clj files are being used as regular namespace/code files (ending with .project? I'd imagine that would be considered an anti-pattern using lein.

The files don't end with `.project` - they end with `.clj`:

```
marc@blaster ~/dev/circleci/circle $ find . -name project.clj -print -exec head -1 {} \;
./project.clj
(defproject circle (try
./src/circle/project.clj
(ns circle.project
./src/circle/http/api/v1/v1_0/project.clj
(ns circle.http.api.v1.v1-0.project
./src/circle/http/api/v1/v1_1/project.clj
(ns circle.http.api.v1.v1-1.project
./src/circle/http/api/v1/shared/project.clj
(ns circle.http.api.v1.shared.project
./src/circle/http/api/v2/insights/project.clj
(ns circle.http.api.v2.insights.project
./src/circle/http/api/v2/project.clj
(ns circle.http.api.v2.project
./src/circle/http/api/v2/entities/project.clj
(ns circle.http.api.v2.entities.project
./src/circle/model/project.clj
(ns circle.model.project
./src/circle/model/api/project.clj
(ns circle.model.api.project
```

Here you can see file starts with `(defproject ...` and the rest start `(ns ...`.

I totally agree that trying to load the deepest project file first makes sense. We have a bunch of git repos that have multiple projects, too.

I was thinking that one way forward would be so that Calva attempts to parse all files named `project.clj` starting with the document directory and ending with the workspace root. If any attempt to parse fails, then Calva will attempt to parse the next file in the sequence, rather than aborting. Calva will only abort the attempt to  start the REPL if it cannot parse any of the files. Currently it aborts on the first failure.

**Comment by bpringe** [comment]
Sorry for being unclear. I meant the namespace ending in `.project` (i.e. `some.namespace.project`).

I think most people would not be using project.clj in this manner, so if Calva simply went to the next project.clj going up in directories, then if someone has an issue in a project.clj they intend to use as a lein project file, Calva would silently go up to run a project at a level the person didn't intent to run, and they might not know there's actually an issue with that initial project.clj.

I suppose though, that would could actually show some message about each project file that fails to be parsed.

**Comment by bpringe** [comment]
Maybe Calva can search the file for a `defproject` and ignore it if one isn't found, and continue the search upward.

**Comment by marcomorain** [comment]
> Maybe Calva can search the file for a `defproject` and ignore it if one isn't found, and continue the search upward.

That would be an awesome fix.

**Comment by PEZ** [comment]
Yes, sounds good. Should be pretty easy too. We don't have to search in a structural way, I think.

Is it something you would like to provide a PR for, @marcomorain ?
